### PR TITLE
🤡 Add TestClient

### LINKED ISSUE #45: Test Client
- [x] Implement a test client (maybe a new repo if requirement changed) ?

### PR DESCRIPTION
**Pull request by aerosadegh**
closes #45

### CODE REVIEW

File: fastcrawler/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+from fastcrawler.engine import AioHttpEngine
+
+
+class TestServer:
+    def __init__(self):
+        self.responses = {
+            "GET": {"message": "GET request received"},
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils.py]: if you can, please use ENUM for http methods, you may also rework engine!

File: fastcrawler/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+    def get_response(self, method, data=None):
+        response = self.responses.get(method)
+        if data:
+            response["data"] = data
+        return response
+
+
+class TestClient(AioHttpEngine):
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils.py]: Never subclass from actual impl in mocking. always use Protocols or ABC instead!

File: fastcrawler/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            "POST": {"message": "POST request received"},
+            "PUT": {"message": "PUT request received"},
+            "DELETE": {"message": "DELETE request received"},
+        }
+
+    def get_response(self, method, data=None):
+        response = self.responses.get(method)
+        if data:
+            response["data"] = data
+        return response
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils.py]: When you're mocking something, there's 3 things you should really respect
1. The mocked version and real version MUST have same input/output. this engine is so different.
2. All method should be mocked, on the interface
3. You should "understand" why you're mocking something. The reason is to actually not "SEND REQUEST" to any server right? that means we should be also mocking responses! what does this mean? it means user should be able to redirect responses to url, say hey I want "/user/{user_id}" to return this HTML. your mock is not supporting that, which make it useless.

File: fastcrawler/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        response = self.server.get_response("GET")
+        return response
+
+    def post(self, url, data):
+        response = self.server.get_response("POST", data=data)
+        return response
+
+    def delete(self, url):
+        response = self.server.get_response("DELETE")
+        return response
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils.py]: As mentioned, I/O is not same as real interface.

File: fastcrawler/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    def put(self, url, data):
+        response = self.server.get_response("PUT", data=data)
+        return response
+
+
+# prepare test
+
+server = TestServer()
+client = TestClient(server)
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils.py]: That is good decoupling. but we don't have that in actual interface right? then respect the actual flow.
If you want to inject responses, then you should:

```python
mock_engine = MockEngine() # this just and only just mock the client
mock_client = MockClient(mapped_routes: dict, mock_engine) # this is high level, for more behaviour unsupported to mock the actual server and routing

router.get(...)
```

File: fastcrawler/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+assert response == {"message": "POST request received", "data": {"key": "value"}}
+
+# Test DELETE method
+response = client.delete("http://example.com")
+assert response == {"message": "DELETE request received"}
+
+# Test PUT method
+data = {"key": "new_value"}
+response = client.put("http://example.com", data=data)
+assert response == {"message": "PUT request received", "data": {"key": "new_value"}}
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils.py]: make sure your mock is responsbile for input data, such as "data". Make sure it actually aware of which HTTP method is sent. you may return 404 if it's not found for instance, or raise error!

File: fastcrawler/test_utils.py (hunk 1 above)
Comment (aerosadegh) [reply fastcrawler/test_utils.py]: there is a same code in the [aiohttp.hdrs](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/hdrs.py)
should we rewrite ourself own ENUM?

Comment (ManiMozaffar) [reply fastcrawler/test_utils.py]: Of course!
we don't care about the engine we're using. It's an abstraction layer, other engine shouldn't be aware that we're using aiohttp enigne

File: fastcrawler/test_utils.py [hunk 7]
```diff
@@ -1,62 +1,73 @@
...
+class MockEngine:
+    def __init__(
+        self,
+        cookies: list[SetCookieParam] | None = None,
+        headers: dict | None = None,
+        user_agent: str | None = None,
+        proxy: ProxySetting | None = None,
+        connection_limit: int | None = None,
+    ):
+        super().__init__(cookies, headers, user_agent, proxy, connection_limit)
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils.py]: what is super()??? you don't inheritance from anything!

File: fastcrawler/test_utils.py [hunk 8]
```diff
@@ -1,62 +1,73 @@
...
+    async def put(self, requests: list[Request]) -> dict[Url, RequestCycle]:
+        """PUT HTTP Method for protocol to crawl a list of URL."""
+        return await self.batch(requests, "PUT")
 
-    def put(self, url, data):
-        response = self.server.get_response("PUT", data=data)
-        return response
+    async def delete(self, requests: list[Request]) -> dict[Url, RequestCycle]:
+        """DELETE HTTP Method for protocol to crawl a list of URL."""
+        return await self.batch(requests, "DELETE")
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils.py]: Now that's some great stuff! remember to include other method in contracts as well. there's one to translate response. if you don't mock everything, then your mock won't work. 
Remember you should also mock batch method, which I don't see in your code.

File: fastcrawler/test_utils.py [hunk 9]
```diff
@@ -1,62 +1,73 @@
...
-data = {"key": "value"}
-response = client.post("http://example.com", data=data)
-assert response == {"message": "POST request received", "data": {"key": "value"}}
+# prepare test
 
-# Test DELETE method
-response = client.delete("http://example.com")
-assert response == {"message": "DELETE request received"}
+mapped_routes = {
+    (Method.GET, Request(url="/user/1")): Response(text=str({"id": 1, "username": "admin"})),
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils.py]: That's not good, few things you missed
1. Response MAY have header
2. Response MAY have cookie
3. Response MUST have status_code
4. Response can be in form of JSON or HTML. (keep that in mind). don't use str(dict[str,Any]), load the json correctly.
5. Same stuff goes for request, they MAY have those. 
6. Be more generic, like /user/{user_id}/, then everything mapped to like user/1/ would be caught by this key. maybe you shouldn't use exactly a dictionary :)

**Review by ManiMozaffar — CHANGES_REQUESTED** [review]

File: fastcrawler/test_utils.py (hunk 7 above)
Comment (aerosadegh) [reply fastcrawler/test_utils.py]: will removed

File: fastcrawler/test_utils.py (hunk 9 above)
Comment (aerosadegh) [reply fastcrawler/test_utils.py]: Overall, this is better than previous.
Also it needs some things missed that you mentioned.

Comment (ManiMozaffar) [reply fastcrawler/test_utils.py]: Yes, true :)

File: fastcrawler/test_utils.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    # Extract the user_id from the request URL
+    user_id = re.search(r"/user/([^/]+)", request.url).group(1)
+    return Response(text=f"User {user_id}", status_code=200)
+
+
+# Create a test server instance with a default response
+test_server = TestServer(default_response=default_response)
+
+# Create a request object with a URL pattern that includes a placeholder
+request = Request(url="/user/{user_id}", method="GET")
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils.py:97]: use RequestPath to be more detailed about the flow of execution

Comment (ManiMozaffar) [inline fastcrawler/test_utils.py]: You may use RequestPath to be more detailed
```python
@dataclass
class RequestPath:
    _types = {"user_id": int}
    path = "{user_id}"
    method = "GET"
    response = ReadFile("/html/request.html")
```

**Review by ManiMozaffar — CHANGES_REQUESTED** [review]

**Comment by ManiMozaffar** [comment]
https://pypi.org/project/http-router/

File: fastcrawler/test_utils_tmp.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    Request,
+    # EngineProto,
+)
+
+
+class Method(StrEnum):
+    GET = "GET"
+    POST = "POST"
+    PUT = "PUT"
+    DELETE = "DELETE"
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils_tmp.py:20]: Add more HTTP methods, like patch and option and etc. check aiohttp library

Comment (ManiMozaffar) [inline fastcrawler/test_utils_tmp.py:20]: Add more HTTP methods, like patch and option and etc. check aiohttp library
ooh i think also inside engine we need them as well -:)

File: fastcrawler/test_utils/test_server.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+
+for test_request in test_requests:
+    # Get the response from the test server
+    test_response = asyncio.run(test_server.get_response(test_request.request))
+    # Verify that the response is as expected
+    print(test_request.request)
+    for attr, val in test_request.attrs.items():
+        assert getattr(test_response, attr) == val
+        print(f"  :: {attr!r} == {val!r} -- Passed!")
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_server.py]: I'm not quite sure why is this here, but i think it should be inside test folder, not inside framework itself.

File: fastcrawler/test_utils_tmp.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+# Create a request object with a specific value for the user_id placeholder
+test_request = Request(url="/user/123", method="GET")
+
+# Get the response from the test server
+test_response = test_server.get_response(test_request)
+
+# Verify that the response is as expected
+assert test_response.status_code == 200
+assert test_response.text == "User 123"
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils_tmp.py]: put your assertion and test inside the test folder

**Review by ManiMozaffar — APPROVED** [review]

File: fastcrawler/test_utils_tmp.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+
+# Create a test server instance with a default response
+test_server = TestServer(default_response=default_response)
+
+# Create a request object with a URL pattern that includes a placeholder
+request = Request(url="/user/{user_id}", method="GET")
+
+# Add the user_details_response to the test server
+test_server.add_response(request, user_details_response)
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils_tmp.py:100]: this is veryyy nice

**Review by ManiMozaffar — CHANGES_REQUESTED** [review]

File: fastcrawler/test_utils/test_server.py (hunk 12 above)
Comment (aerosadegh) [reply fastcrawler/test_utils/test_server.py]: We need TestServer, But we will write tests from line 26 to the end in the test folder later!

File: fastcrawler/test_utils_tmp.py (hunk 13 above)
Comment (aerosadegh) [reply fastcrawler/test_utils_tmp.py]: This file will merge into test_utils and test folder
This is temporary!

File: fastcrawler/test_utils/endpoint.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+async def not_allowed(*args, headers, text="Method Not Allowed", **kwargs):
+    return Response(text=text, status_code=405, headers=headers)
+
+
+class HTTPEndpoint:
+    def __init__(self) -> None:
+        self._allowed_methods = [
+            method
+            for method in ("GET", "HEAD", "POST", "PUT", "PATCH", "DELETE", "OPTIONS")
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoint.py:17]: Can you use Enum please? :)

Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoint.py]: Can you use Enum please? :)

File: fastcrawler/test_utils/endpoint.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+        content = await self.get_content()
+        return Response(text=content, **self.response_kwargs)
+
+    async def get_content(self) -> str:
+        if isinstance(self.file, os.PathLike):
+            # Convert the PathLike object to a str
+            file_name = str(self.file)
+        else:
+            # Assume the file is a str
+            file_name = self.file
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoint.py]: This isn't type safe!
This should never be executed, your IDE should also show you warning.
Try to parse it, not writing if/else. Maybe you can add a utilty function that converts anything else to os.PathLike

File: fastcrawler/test_utils/endpoint.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+        # Check if the file name is valid and is an html file
+        if self.is_valid_file_name(file_name):
+            async with aiofiles.open(file_name, "r") as f:
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoint.py]: as "file". try to be more explicit when naming variable, especially as we're writing an open source tool

File: fastcrawler/test_utils/endpoint.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+                content = await f.read()
+            return content
+        else:
+            raise ValueError("Invalid file name or file type")
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoint.py]: instead of having a very long if / else, just negative the first condition and then raise error. if not is valid file name, then raise error. then your code would be less nested and better :)

File: fastcrawler/test_utils/endpoint.py [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+    def is_valid_file_name(self, file_name: str) -> bool:
+        return os.path.exists(file_name) and file_name.endswith(".html")
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoint.py]: Don't validate, try to parse this. Maybe you can create a new string type, that only get parsed if these two condition meets, like HTMLFile(...)

File: fastcrawler/test_utils/routes.py [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+        match = self.path_regex.match(url)
+        path_params = {}
+        if match:
+            matched_params = match.groupdict()
+            for key, value in matched_params.items():
+                matched_params[key] = self.param_convertors[key].convert(value)
+            path_params.update(matched_params)
+            method_func = self.response.dispatch(method.lower())
+            return method_func(url, params=path_params)
+        raise NoMatchFound(url, path_params)
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/routes.py]: again raise error on top, saying if not match, then raise it.

File: fastcrawler/test_utils/test_server.py [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+    def add_route(self, route: Route):
+        self.routes.append(route)
+
+    def get_response(self, request: Request) -> Response:
+        for route in self.routes:
+            try:
+                return route.url_path_for(request.url, request.method)
+            except NoMatchFound:
+                pass
+        raise NoMatchFound(request.url, {})
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_server.py]: I don't understand this function
You shouldn't do dummy raising, raise the exception when it happened, there's no need to perform all the loop then raise the exception. it just makes the debugging much harder.

File: test/conftest.py [hunk 22]
```diff
@@ -72,4 +74,14 @@ def task_manager(task_app):
...
+
+
+@pytest.fixture(scope="session")
+def test_server():
+    yield TestServer(routes=routes)
+
+
+@pytest.fixture(scope="session")
+def test_requests():
+    yield __test_requests
```
Comment (ManiMozaffar) [inline test/conftest.py]: don't define a fixture that is only used in one test, inside conftest. conftest is place for global fixture, which would be used all around the tests.

File: test/shared/test_server.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+        return Response(text=text, status_code=200)
+
+
+def read_html_file(html_file):
+    with open(html_file) as html:
+        html_content = html.read()
+    return html_content
+
+
+html_file = Path(__file__).parent / "sample_html_file.html"
```
Comment (ManiMozaffar) [inline test/shared/test_server.py]: again use HTMLFile object (design it) to be more readable

File: test/shared/test_server.py [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+        text = "Hello"
+        if user_id:
+            text += f", {user_id}"
+        if post_id:
+            text += f" in {post_id}"
+        if allow:
+            text += f", allow is {allow}"
+        if query:
+            text += f", query for {query}"
+        return Response(text=text, status_code=200)
```
Comment (ManiMozaffar) [inline test/shared/test_server.py]: While this is good and very flexible, but i'd also want to have a builder pattern here
ResponseBuilder(status=200, methods="GET")
this way, while user have the ability to be flexible, he can also be fast at mocking APIs.

File: test/test_test_server.py [hunk 25]
```diff
@@ -0,0 +1,11 @@
...
+
+
+@pytest.mark.asyncio
+async def test_test_server_response(test_server, test_requests):
+    for test_request in test_requests:
+        # Get the response from the test server
+        test_response = await test_server.get_response(test_request.request)
+        # Verify that the response is as expected
+        for attr, val in test_request.attrs.items():
+            assert getattr(test_response, attr) == val
```
Comment (ManiMozaffar) [inline test/test_test_server.py]: 1. Why there's only one test?
2. The test is not testing the "mocking" behavior. Try to actually write a end to end test that checks that our mocked class is working as expected. for that, you can create an API on our FastAPI application, then mock it inside engine, and then write a test that send request to there, and then request to fastapi, and assert that everything inside the Response object is same.

**Review by ManiMozaffar — CHANGES_REQUESTED** [review]

File: test/shared/test_server.py (hunk 24 above)
Comment (aerosadegh) [reply test/shared/test_server.py]: We can write a ResponseBuilder class like that :)
It's very like to HTMLResponse

File: test/test_test_server.py (hunk 25 above)
Comment (aerosadegh) [reply test/test_test_server.py]: yes
It's just test the test_server
we will write MockEngine in the next!

File: fastcrawler/test_utils/test_server.py (hunk 21 above)
Comment (aerosadegh) [reply fastcrawler/test_utils/test_server.py]: This is a searching in the routes to find the matched function for the method and the url.
If there is no match(End of for), so the passed url and method not defined in the server and raise error!

**Comment by ManiMozaffar** [comment]
```python
server = MockServer() # mock all urls that he wants to send request to
engine = MockEngine(server=server)
fastcrawler.override_dependencies[get_engine] = lambda engine
# TODO: digikala.com/... -> Mock response -> Server
```

File: fastcrawler/core/spider.py [hunk 26]
```diff
@@ -44,15 +44,15 @@ def __init__(
...
         if model is None and self.data_model is None:
-            raise SpiderBadConfiguratonError("A data model must be specified")
+            raise SpiderBadConfigurationError("A data model must be specified")
 
         if self.engine_request_limit is None and self._engine.default_request_limit is None:
-            raise SpiderBadConfiguratonError("Request limit must be specified")
+            raise SpiderBadConfigurationError("Request limit must be specified")
 
         self.parser = parser or HTMLParser
         self._data_model: Type[BaseModel] = model or self.data_model  # type: ignore
```
Comment (ManiMozaffar) [inline fastcrawler/core/spider.py:53]: why did you write type ignore?

File: fastcrawler/test_utils/endpoints.py [hunk 27]
```diff
@@ -0,0 +1,182 @@
...
+
+    Attributes:
+        _allowed_methods (list[str]): A list of allowed HTTP methods for this endpoint.
+    """
+
+    def __init__(self) -> None:
+        """Initialize the base endpoint with the allowed methods."""
+        self._allowed_methods = [
+            method
+            for method in [str(member) for member in HTTPMethod.__members__.values()]
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoints.py]: instead of doing HTTPMethod.__members__.values, do in(member.value) for member in HTTPMethod

this is more readable :)

File: fastcrawler/test_utils/endpoints.py [hunk 28]
```diff
@@ -0,0 +1,182 @@
...
+    def dispatch(self, request_method: str) -> Any:
+        """Dispatch the request to the appropriate method.
+
+        Args:
+            request_method (str): The request method to dispatch.
+
+        Returns:
+            Any: The result of calling the corresponding method.
+        """
+        if resp := self.method_not_allowed(request_method):
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoints.py:89]: Any reason you decide to us walrus operator here?

File: fastcrawler/test_utils/endpoints.py [hunk 29]
```diff
@@ -0,0 +1,182 @@
...
+
+    def __init__(self) -> None:
+        """Initialize the base endpoint with the allowed methods."""
+        self._allowed_methods = [
+            method
+            for method in [str(member) for member in HTTPMethod.__members__.values()]
+            if getattr(self, method.lower(), None) is not None
+        ]
+
+    def dispatch(self, request_method: str) -> Any:
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoints.py]: be more specific in typing hinting, Any shouldn't be used.

File: fastcrawler/test_utils/endpoints.py [hunk 30]
```diff
@@ -0,0 +1,182 @@
...
+
+        Args:
+            method (Union[HTTPMethod, str]): The HTTP method to handle.
+            **response_kwargs: The keyword arguments to pass to the response object.
+
+        Raises:
+            ValueError: If the method is not a member of HTTPMethod.
+        """
+
+        async def http_method(self, *args, **kwargs) -> Response:
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoints.py]: why have you used async function here?

File: fastcrawler/test_utils/routes.py [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+
+        Args:
+            request_pattern (str): The request pattern to match.
+            response (BaseEndpoint): The response object to handle the request.
+        """
+        self.request_pattern = request_pattern
+        self.response = response
+        self.path_regex, self.path_format, self.param_convertors = compile_path(request_pattern)
+
+    def url_path_for(self, url: str, method: str) -> Callable[..., Response]:
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/routes.py]: rename this function to something more meaningful
maybe get_response_from_url

File: fastcrawler/test_utils/endpoints.py [hunk 32]
```diff
@@ -0,0 +1,182 @@
...
+
+        Args:
+            html_file (HTMLPathType): The HTML file path to read from.
+            **response_kwargs: The keyword arguments to pass to the response object.
+        """
+        super().__init__()
+        self.html_file = HTMLPath(html_file)
+        self.response_kwargs = response_kwargs
+
+    async def get(self, *args, **kwargs) -> Response:
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/endpoints.py]: I'm confused here
you only have GET, how do you handle POST and ETC? do you actually serve it with a GET request? with if request has body?

File: fastcrawler/test_utils/test_engine.py [hunk 33]
```diff
@@ -0,0 +1,108 @@
...
+class MockEngine:
+    """This class uses the TestServer instance to handle requests and return responses.
+
+    Attributes:
+        servers (TestServer | Iterable[TestServer]): The test server instances.
+        base_urls (str | Iterable[str]): The base urls of the requests.
+    """
+
+    def __init__(
+        self, servers: TestServer | Iterable[TestServer], base_urls: str | Iterable[str]
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_engine.py]: servers: TestServer | Iterable[TestServer]
why is this Iterable of testserver, and testserver itself at the same time? wouldn't it be more meaningful if you just get the Iterable? and don't be too much flexible

Comment (ManiMozaffar) [reply fastcrawler/test_utils/test_engine.py]: same goes for base_urls

File: fastcrawler/test_utils/test_engine.py [hunk 34]
```diff
@@ -0,0 +1,108 @@
...
+    ) -> None:
+        servers = servers if not isinstance(servers, TestServer) else [servers]
+        base_urls = base_urls if not isinstance(base_urls, str) else [base_urls]
+
+        self.servers = {base_url: server for base_url, server in zip(base_urls, servers)}
+
+    @staticmethod
+    def _extract_base_url_and_route(url: str) -> tuple[str, str]:
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_engine.py]: tuple[str, str] 
doesn't say much about what's being returned

File: fastcrawler/test_utils/test_engine.py [hunk 35]
```diff
@@ -0,0 +1,108 @@
...
+    """This class uses the TestServer instance to handle requests and return responses.
+
+    Attributes:
+        servers (TestServer | Iterable[TestServer]): The test server instances.
+        base_urls (str | Iterable[str]): The base urls of the requests.
+    """
+
+    def __init__(
+        self, servers: TestServer | Iterable[TestServer], base_urls: str | Iterable[str]
+    ) -> None:
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_engine.py]: You don't need type anno on __init__

File: fastcrawler/test_utils/test_server.py [hunk 36]
```diff
@@ -0,0 +1,64 @@
...
+class TestServer:
+    """A class that represents a test server.
+
+    This class can store a list of routes and return a response for a given request.
+
+    Attributes:
+        routes (list[Route]): A list of routes that the server can handle.
+    """
+
+    def __init__(self, routes: Optional[list[Route]] = None) -> None:
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_server.py]: Don't use Optional. list[Route] | None

File: fastcrawler/test_utils/test_server.py [hunk 37]
```diff
@@ -0,0 +1,64 @@
...
+            Optional[Route]: The matching route or None if not found.
+        """
+        # Loop through the routes and return the first response that matches
+        for route in self.routes:
+            try:
+                return route.url_path_for(url, method)
+            except NoMatchFound:
+                pass
+        # If no route matches, return None
+        return None
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_server.py]: why dummy return? just keep raising the exception

File: fastcrawler/test_utils/test_server.py [hunk 38]
```diff
@@ -0,0 +1,64 @@
...
+
+        Returns:
+            Optional[Route]: The matching route or None if not found.
+        """
+        # Loop through the routes and return the first response that matches
+        for route in self.routes:
+            try:
+                return route.url_path_for(url, method)
+            except NoMatchFound:
+                pass
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_server.py]: rework this with first_true :)
from more_itertools library.

File: fastcrawler/test_utils_tmp.py [hunk 39]
```diff
@@ -0,0 +1,110 @@
...
+    ):
+        ...
+
+    async def batch(
+        self,
+        requests: list[Request],
+        method: str,
+    ) -> dict[Url, RequestCycle]:  # type: ignore
+        """Batch Method for protocol to retrieve a list of URL."""
+        return ...
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils_tmp.py]: is this working?

File: fastcrawler/test_utils_tmp.py (hunk 13 above)
Comment (ManiMozaffar) [reply fastcrawler/test_utils_tmp.py:110]: this is still not moved to test directory

File: fastcrawler/utils/injection.py [hunk 40]
```diff
@@ -47,36 +47,35 @@ def __repr__(self) -> str:
...
     @staticmethod
     async def inject(func: "_Depends"):
+        sig = inspect.signature(func.dependency)
+        bound = sig.bind_partial(*func.args, **func.kwargs)
+        bound.apply_defaults()
+
+        for name, value in bound.arguments.items():
+            if isinstance(value, _Depends):
+                bound.arguments[name] = await _Depends.evaluate_dependency(value)
+
```
Comment (ManiMozaffar) [inline fastcrawler/utils/injection.py:57]: why did you change the codes here? any reason?

File: test/test_test_server.py [hunk 41]
```diff
@@ -0,0 +1,55 @@
...
+    # Get the response from the test server
+    url_test_response_dct = await mock_engine.batch(
+        [test_request.http_request for test_request in test_requests]
+    )
+    # Verify that the response is as expected
+    for test_request in test_requests:
+        test_response = url_test_response_dct[test_request.http_request.url]
+        for exp_attr, exp_val in test_request.expected_response_attributes.items():
+            assert hasattr(test_response, exp_attr)
+            assert getattr(test_response, exp_attr) == exp_val
```
Comment (ManiMozaffar) [inline test/test_test_server.py]: these 2 tests are fine, but:
1. I'd write more tests for edge cases. Maybe you could have included json response as well, rather than just HTML
2. Maybe create a HTTP app, and compare it with your mock? to make sure that your mock is working same as the real object. We normally don't write test to confirm that our "mock" is mocking correctly, but here in this context we must, because other people are also gonna use this mock. that's what i mean:
send req to test://some_url/some_path
send req to localhost:8000/some_path
compare them, and check if the result is as expected.

This would give us more confidence about our mock :)

**Review by ManiMozaffar — CHANGES_REQUESTED** [review]

File: fastcrawler/core/spider.py (hunk 26 above)
Comment (aerosadegh) [reply fastcrawler/core/spider.py:53]: mypy error:
Incompatible types in assignment (expression has type "type[BaseModel] | None", variable has type "type[BaseModel]")  [assignment]

File: fastcrawler/test_utils/endpoints.py (hunk 28 above)
Comment (aerosadegh) [reply fastcrawler/test_utils/endpoints.py:89]: any alternate way to write this block?
(walrus operator added in python 3.8)

File: fastcrawler/test_utils/endpoints.py (hunk 30 above)
Comment (aerosadegh) [reply fastcrawler/test_utils/endpoints.py]: the async expression removed

File: fastcrawler/test_utils/endpoints.py (hunk 32 above)
Comment (aerosadegh) [reply fastcrawler/test_utils/endpoints.py]: we write this class for handle html files
```python
class StaticResponse(BaseEndpoint):
    """A class that represents a static HTML response...."""
``` 
if the name isn't meaningful so we most change the name.

we can POST or etc methods in html file?

File: fastcrawler/test_utils/test_engine.py (hunk 33 above)
Comment (aerosadegh) [reply fastcrawler/test_utils/test_engine.py]: Shouldn't we be kind to the user? :)

ok it's fixed!

File: fastcrawler/test_utils/test_server.py (hunk 38 above)
Comment (aerosadegh) [reply fastcrawler/test_utils/test_server.py]: it's forces the new dependency, we have three way:
1. use this without any change
2. use more_itertools and force new dependency
3. rewrite first_true in this project

File: fastcrawler/utils/injection.py (hunk 40 above)
Comment (aerosadegh) [reply fastcrawler/utils/injection.py:57]: To reduce mypy errors and write in better way (it seems has been duplicate code in the algorithm)
This part of code not related to this pr.
we can revert this part

File: fastcrawler/test_utils_tmp.py (hunk 39 above)
Comment (aerosadegh) [reply fastcrawler/test_utils_tmp.py]: It's removed

File: fastcrawler/test_utils/endpoints.py (hunk 32 above)
Comment (aerosadegh) [reply fastcrawler/test_utils/endpoints.py]: We changed this part to handle also JSON files, so we able to use all http methods to this class :)

**Review by amirbahador-hub — APPROVED** [review]

File: fastcrawler/test_utils/test_engine.py [hunk 42]
```diff
@@ -1,15 +1,43 @@
...
+            UrlInfo: A tuple of the base URL and the route.
+        """
+        match = _url_pattern.match(url)
+        if match:
+            base_url, route = match.groups()
+            return cls(base_url, route or "/")
+
+        # If the URL does not match regex, set the base URL to an empty string
+        # and the route to the same as the URL
+        return cls("", url)
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_engine.py]: Pydantic has a URL Type, rest of PR looks fine to me!

**Review by ManiMozaffar — CHANGES_REQUESTED** [review]

File: fastcrawler/test_utils/test_server.py (hunk 38 above)
Comment (ManiMozaffar) [reply fastcrawler/test_utils/test_server.py]: more_itertools is not a "heavy" dependency, 
I'd argue we can have it, we may need it in future as well.
and indeed, algorithms used there is much better than when we write s.th by ourself quick.

File: fastcrawler/test_utils/test_engine.py (hunk 33 above)
Comment (ManiMozaffar) [reply fastcrawler/test_utils/test_engine.py]: > Shouldn't we be kind to the user? :)

No. I did same mistake in codebase. i will fix my own codes as well in another PR when i'm free.

File: fastcrawler/core/spider.py (hunk 26 above)
Comment (ManiMozaffar) [reply fastcrawler/core/spider.py:53]: ok i'll check and fix mypy errors in another PR, whole codebase has to adopt it at end.

File: fastcrawler/test_utils/test_server.py (hunk 38 above)
Comment (aerosadegh) [reply fastcrawler/test_utils/test_server.py]: The `more_itertools` is added!

File: fastcrawler/test_utils/test_server.py [hunk 43]
```diff
@@ -0,0 +1,67 @@
...
+
+        Raises:
+            NoMatchFound: If no route matches the request.
+        """
+        # Use more_itertools.first_true to find the first response that matches
+        matched_route = first_true(
+            self.routes,
+            None,
+            lambda route: route.get_response_from_url(url, method, raise_exception=False),
+        )
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_server.py]:
```
def prediction(route: Route):
     try:
            return s.th
     except:
            return None

matched_route = first_true(
            self.routes,
            None,
            prediction,
        )
 ```

**Review by ManiMozaffar — COMMENTED** [review]
noticed s.th

File: fastcrawler/test_utils/test_server.py [hunk 44]
```diff
@@ -50,17 +51,26 @@ def _find_response_function(self, url: str, method: str) -> Callable[..., Respon
...
             NoMatchFound: If no route matches the request.
         """
+
         # Use more_itertools.first_true to find the first response that matches
+        def prediction(route: Route, url, method):
+            try:
+                route.get_response_from_url(url, method)
+                return True
+            except NoMatchFound:
+                pass
```
Comment (ManiMozaffar) [inline fastcrawler/test_utils/test_server.py:66]: that's nice ;)

**Review by ManiMozaffar — APPROVED** [review]
LGTM
